redis_session.c

```diff
@@ -183,6 +186,130 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+
+        calculate_secret_hash(lock_status);
+
+        lock_wait_time = INI_INT("redis.session.lock_wait_time");
+        if (lock_wait_time == 0) {
+          lock_wait_time = 2000;
+        }
+
+        max_lock_retries = INI_INT("redis.session.lock_retries");
+        if (max_lock_retries == 0) {
+          max_lock_retries = 10;
+        }
+
+        lock_expire = INI_INT("redis.session.lock_expire");
+        if (lock_expire == 0) {
+          lock_expire = INI_INT("max_execution_time");
+        }
+
+        // Building the redis lock key
+        smart_string_appendl(&lock_key, lock_status->session_key, strlen(lock_status->session_key));
+        smart_string_appendl(&lock_key, "_LOCK", strlen("_LOCK"));
+        smart_string_0(&lock_key);
+        strncpy(lock_status->lock_key, lock_key.c, 256);
+
+        if (lock_expire > 0) {
+            cmd_len = REDIS_SPPRINTF(&cmd, "SET", "ssssd", lock_status->lock_key, strlen(lock_status->lock_key), lock_status->lock_secret_hash, strlen(lock_status->lock_secret_hash), "NX", 2, "PX", 2, lock_expire * 1000);
+        } else {
+            cmd_len = REDIS_SPPRINTF(&cmd, "SET", "sss", lock_status->lock_key, strlen(lock_status->lock_key), lock_status->lock_secret_hash, strlen(lock_status->lock_secret_hash), "NX", 2);
+        }
+
+        for (i_lock_retry = 0; !lock_status->is_locked && (max_lock_retries == -1 || i_lock_retry <= max_lock_retries); i_lock_retry++) {
+          if(!(redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC) < 0)
+              && ((response = redis_sock_read(redis_sock, &response_len TSRMLS_CC)) != NULL)
+              && response_len == 3
+              && strncmp(response, "+OK", 3) == 0) {
+                lock_status->is_locked = 1;
+          } else if (max_lock_retries == -1 || i_lock_retry < max_lock_retries) {
+              usleep(lock_wait_time);
+          }
+        }
+        lock_status->is_locked = 1;
+
+        efree(cmd);
+        efree(response);
+    }
+}
+
+void lock_release(RedisSock *redis_sock, redis_session_lock_status *lock_status)
+{
+    if (lock_status->is_locked) {
+        char *cmd, *response;
+        int response_len, cmd_len;
+
+        upload_lock_release_script(redis_sock);
+        cmd_len = REDIS_SPPRINTF(&cmd, "EVALSHA", "sdss", REDIS_G(lock_release_lua_script_hash), strlen(REDIS_G(lock_release_lua_script_hash)), 1, lock_status->lock_key, strlen(lock_status->lock_key), lock_status->lock_secret_hash, strlen(lock_status->lock_secret_hash));
+
+        redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+        response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
+
+        // in case of redis script cache has been flushed
+        if (response == NULL) {
+            REDIS_G(lock_release_lua_script_uploaded) = 0;
+            upload_lock_release_script(redis_sock);
+            redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+            response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
```

Comment:
Thank you for the hint. In this particular case the response is not used.
But in other cases, as for example LUA script upload, we do. How would you suggest should we handle errors like NULL return there (Warning, failure return, etc)?